torch/_subclasses/functional_tensor.py

```diff
@@ -331,41 +340,31 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+            # Never decompose dropout in export
             if self.export and func == torch.ops.aten.dropout.default:
                 return False
-            # TODO (tmanlaibaatar)
-            # Eventually, we don't want to decompose any aten op at all
-            # but there is a safety and coverage gap that we need to close
-            # before that.
-            #
-            # (1) the "safety" is what we are risking with this PR
-            #     (we are blindly taking every op that advertises as
-            #      functional and sending it to the functional fallback.
-            #      We risk silent correctness if we have an op that lies about its schema,
-            #      that we didn't manually hardcode above) Therefore we always decompose them
-            # (2) the "not every composite inplace op has a functional variant" is a coverage gap,
-            #      but not really a safety risk, since we'll loudly error when we try to generate
-            #      functionalization kernels for these new (composite) inplace/view ops. But until we
-            #      establish such gap more concretely, we still decompose them
-            if self._dispatch_key is not None:
-                # it is unsafe to not decompose ops that claim to be functional but actually aren't
-                if func in FunctionalTensor.maybe_aliasing_or_mutating_ops:
-                    return True
-                # only decompose view or inplace mutating ops
-                alias_info = len(
-                    [i for i in func._schema.arguments if i.alias_info is not None]
-                )
-                should_decompose = alias_info != 0 or func._schema.is_mutable
-                if not should_decompose:
-                    if func.namespace not in ["aten", "prim"]:
-                        warnings.warn(
-                            f"At pre-dispatch tracing, we will assume that any "
-                            f"custom op that is marked with CompositeImplicitAutograd "
-                            f"and functional are safe to not decompose. We found {func}"
-                            f" to be one such op."
-                        )
-                return should_decompose
+
+            # We unconditionally decompose ops that are maybe aliasing or mutating ops
+            if func in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+                return True
+
+            # We unconditionally decompose ops that are mutating or aliasing ops
+            alias_info_present = any(arg.alias_info for arg in func._schema.arguments)
+            if alias_info_present or func._schema.is_mutable:
+                return True
+
+            # If we are here, it means we are seeing functional composite op.
+            # For pre-dispatch IR or export inference IR, we wont' decompose them
+            if self.export or self.pre_dispatch:
+                if func.namespace not in ["aten", "prim"]:
+                    warnings.warn(
```

Comment:
hmmm.... this warning doesn't feel very actionable

here's a tentative alternative proposal: if the only time we can end up here is if we hit a custom op, then:

(1) if the custom op is pt2-compliant (marked with the `Tag.pt2_compliant` tag), then it promises to be traceable so we should leave it in

(2) if the custom op is not pt2-compliant, then we should error loudly and tell the user to make their op compliant. If dynamo is running (via strict export) we should never end up here, since dynamo will have graph broken / error'd on the op.

Reply:
ok sounds good!

Reply:
Will do as follow up